application/src/main/java/org/thingsboard/server/actors/device/DeviceActorMessageProcessor.java

```diff
@@ -588,34 +625,38 @@ private void processRpcResponses(SessionInfoProto sessionInfo, ToDeviceRpcRespon
         log.debug("[{}][{}] Processing RPC command response: {}", deviceId, sessionId, responseMsg);
         int requestId = responseMsg.getRequestId();
         ToDeviceRpcRequestMetadata requestMd = toDeviceRpcPendingMap.remove(requestId);
-        boolean success = requestMd != null;
-        if (success) {
-            ToDeviceRpcRequest toDeviceRequestMsg = requestMd.getMsg().getMsg();
-            boolean delivered = requestMd.isDelivered();
-            boolean hasError = StringUtils.isNotEmpty(responseMsg.getError());
-            try {
-                String payload = hasError ? responseMsg.getError() : responseMsg.getPayload();
-                systemContext.getTbCoreDeviceRpcService().processRpcResponseFromDeviceActor(
-                        new FromDeviceRpcResponse(toDeviceRequestMsg.getId(), payload, null));
-                if (toDeviceRequestMsg.isPersisted()) {
-                    RpcStatus status = hasError ? RpcStatus.FAILED : RpcStatus.SUCCESSFUL;
-                    JsonNode response;
-                    try {
-                        response = JacksonUtil.toJsonNode(payload);
-                    } catch (IllegalArgumentException e) {
-                        response = JacksonUtil.newObjectNode().put("error", payload);
-                    }
-                    systemContext.getTbRpcService().save(tenantId, new RpcId(toDeviceRequestMsg.getId()), status, response);
-                }
-            } finally {
-                if (!delivered) {
-                    String errorResponse = hasError ? "error" : "";
-                    log.debug("[{}][{}][{}] Received {} response for undelivered RPC! Going to send next pending request ...", deviceId, sessionId, requestId, errorResponse);
-                    sendNextPendingRequest();
+        if (requestMd == null) {
+            log.debug("[{}][{}][{}] RPC command response is stale!", deviceId, sessionId, requestId);
+            return;
+        }
+        ToDeviceRpcRequest toDeviceRequestMsg = requestMd.getMsg().getMsg();
+        UUID rpcId = toDeviceRequestMsg.getId();
+        boolean delivered = requestMd.isDelivered();
+        boolean hasError = StringUtils.isNotEmpty(responseMsg.getError());
+        try {
+            String payload = hasError ? responseMsg.getError() : responseMsg.getPayload();
```

Comment:
@YevhenBondarenko, in common scenarios, we don't have an error in the `ToDeviceRpcResponseMsg`. However, if the error happened, for example, an error on the transport side during a request sent to the device we added as an error a string value. For example:

```
String error = "Failed to convert device RPC command to MQTT msg: " + rpcRequest.getMethodName() + rpcRequest.getParams();
```
We are passing this a response to the API and Rule Engine. On the API side:

```
Optional<String> responseData = response.getResponse();
if (responseData.isPresent() && !StringUtils.isEmpty(responseData.get())) {
    String data = responseData.get();
    try {
        logRpcCall(rpcRequest, rpcError, null);
        responseWriter.setResult(new ResponseEntity<>(JacksonUtil.toJsonNode(data), HttpStatus.OK));
    } catch (IllegalArgumentException e) {
        log.debug("Failed to decode device response: {}", data, e);
        logRpcCall(rpcRequest, rpcError, e);
        responseWriter.setResult(new ResponseEntity<>(HttpStatus.NOT_ACCEPTABLE));
    }
} else {
    logRpcCall(rpcRequest, rpcError, null);
    responseWriter.setResult(new ResponseEntity<>(HttpStatus.OK));
}
```

`JacksonUtil.toJsonNode(data)` will always fail. As a result we will always return `responseWriter.setResult(new ResponseEntity<>(HttpStatus.NOT_ACCEPTABLE));`

Similar issue on the Rule Engine side. The issue is that our response will be always a string instead of json object.